keps/sig-network/3726-standard-application-protocols/README.md

```diff
@@ -0,0 +1,236 @@
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+There are cases where implementations implement different things for the same application protocol names. That already causes(and can cause more in the future) issues for certain implementations to interoperate. (see example with GKE and Istio under [Motivation](#motivation))
+
+This KEP proposes to declare standard Kubernetes protocol names for common protocols that are not IANA service names.
+
+
+## Motivation
+
+The lack of direct support for specifying application protocols for ports and the widespread use of implementation-specific annotations to mitigate it had led k8s to add the `AppProtocol` field to the port spec.
+
+While this is a good solution - we never came with recommended standards other than [IANA standard service names](https://www.iana.org/assignments/service-names) or a domain prefixed protocol.
+
+This loose definition has led us to have instances where implementations do different things for common protocols.
+
+One good example is `HTTP2`. 
+* In GKE you can use `appProtocol: HTTP2` and it will describe HTTP2 over TLS (https://cloud.google.com/kubernetes-engine/docs/how-to/secure-gateway#load-balancer-tls).
+* While in Istio it will be h2c (HTTP2 over cleartext).
+
+That creates a problem where users with GKE and Istio in their cluster can have very different behaviors for the same `appProtocol` value.
+
+
+
+### Goals
+
+* Build consensus around how common (non IANA service names) should be implemented
+* Help the broader community and specifically implementations to interoperate better
+* Provide short and clear documentation for how appProtocol values should be implemented
+  * Update the appProtocol user documentation respectively
+
+
+### Non-Goals
+
+* Validate appProtocol values
+* Monitor `appProtocol implementations
+
+
+## Proposal
+
+Kubernetes supports the `appProtocol` field to provide a way to specify an application protocol for each Service port.
+
+The [documentation](https://kubernetes.io/docs/concepts/services-networking/service/#application-protocol) for this field says that:
+
+```Values should either be IANA standard service names or domain prefixed names such as mycompany.com/my-custom-protocol.```
+
+This KEP proposes to declare standard Kubernetes protocol names for common protocols that are not IANA standard service names.
+
+Those common protocols will be well defined strings prefixed with ‘k8s.io’. 
+`k8s.io/h2c` as an example.
+
+### list of protocols
+- 'k8s.io/h2c'
+- 'k8s.io/h2'
+- 'k8s.io/grpc'
+- 'k8s.io/grpc-web'
+- 'k8s.io/tls'
+- 'k8s.io/tcp'
```

Comment:
Do we think all of them are needed?

Alternatively, are there protocols I missed?

Reply:
I think I'd drop `grpc-web` (not prevalent enough), and `tcp` (duplicative of port protocol). I'm also not sure how `tls` would differ in practice from `https` that's already defined as an IANA standard name, so I think it could make sense to leave that off as well.

Reply:
`tcp` is not duplicative.


HTTP:
```
appProtocol: k8s.io/HTTP
protocol: TCP
```

"Who knows, do whatever you want":
```
protocol: TCP
```


"This is definitely TCP"
```
appProtocol: k8s.io/TCP #might be IANA though, not sure prefix is needed
protocol: TCP
```

I know at least Istio and Linkerd treat these 3 cases differently

Reply:
You mean that here Istio tries to sniff and guess the app protocol 
```
protocol: TCP
```

and here it does not? 
```
appProtocol: k8s.io/TCP # it is unprefixed "tcp" in the current istio support
protocol: TCP
```

Reply:
Yes. if the duplicative "TCP" is confusing, you can also think of it as 

```
appProtocol: k8s.io/opaque-bytestream # it is unprefixed "tcp" in the current istio support
protocol: TCP
```